src/utils/workspace.ts

```diff
@@ -51,22 +51,30 @@ export async function selectWorkspaceItem(placeHolder: string, options: vscode.O
     return folder && folder.data ? folder.data : (await ext.ui.showOpenDialog(options))[0].fsPath;
 }
 
-export async function showWorkspaceFoldersQuickPick(placeHolderString: string, context: IActionContext, subPathSetting: string | undefined): Promise<string> {
-    context.telemetry.properties.cancelStep = 'showWorkspaceFolders';
+export async function showWorkspaceFoldersAndFiles(placeHolderString: string, context: IActionContext, subPathSetting: string | undefined, fileExtensions?: string | string[]): Promise<string> {
+    context.telemetry.properties.cancelStep = 'showWorkspaceFoldersAndExtensions';
+
+    if (!fileExtensions) {
+        // these are the current deployables we support
+        fileExtensions = ['jar', 'war', 'zip'];
+    } else if (typeof fileExtensions === 'string') {
+        fileExtensions = [fileExtensions];
+    }
+
     return await selectWorkspaceItem(
         placeHolderString,
         {
-            canSelectFiles: false,
-            canSelectFolders: true,
+            canSelectFiles: true,
+            canSelectFolders: false,
```

Comment:
The point of doing `showWorkspaceFoldersAndFiles` was to support both java and non-java in the same code path. If you can't select both files and folders I think that defeats the point

Reply:
So this is true and this was a mistake, buuuuut...

```
        /**
	 * Options to configure the behaviour of a file open dialog.
	 *
	 * * Note 1: A dialog can select files, folders, or both. This is not true for Windows
	 * which enforces to open either files or folder, but *not both*.
	 * * Note 2: Explicitly setting `canSelectFiles` and `canSelectFolders` to `false` is futile
	 * and the editor then silently adjusts the options to select files.
	 */```

😬

Reply:
Summarizing offline discussion:
That's unfortunate. At the end of the day this PR is about improving the experience for the majority of languages by prompting for folder before the web app. We still have to handle edge-cases (aka Java), but it doesn't have to be an ideal user experience. I don't have an exact answer on what the experience should be and I'm hoping you can come up with something 🙂 but let me or Matt know if you need help brainstorming.